File: src/processors/inviteLink/inviteLink.ts
import * as parseUri from 'parse-uri';

import { SafetyJim, MessageProcessor } from '../../safetyjim/safetyjim';
import { Shard } from '../../safetyjim/shard';
import * as Utils from '../../safetyjim/utils';
import { Message, MessageReaction, User, PermissionString } from 'discord.js';
import { Settings } from '../../database/models/Settings';

class InviteLink implements MessageProcessor {
    private blacklistedHosts: string[];
    private shortestHostLength: number;
    private whitelistedRoles: PermissionString[];

    constructor() {
        this.blacklistedHosts = [ 'discord.gg' ];
        this.shortestHostLength = 10;
        this.whitelistedRoles = [
            'ADMINISTRATOR',
            'BAN_MEMBERS',
            'KICK_MEMBERS',
            'MANAGE_ROLES',
            'MANAGE_MESSAGES',
        ];
    }

    public async onMessage(shard: Shard, jim: SafetyJim, msg: Message): Promise<boolean> {
        let user = await shard.client.fetchUser(msg.author.id, true);
        let member = await msg.guild.fetchMember(msg.author.id);

        for (let role of this.whitelistedRoles) {
            if (member.hasPermission(role)) {
                return;
            }
        }

        let enabled = await jim.database.getGuildSetting(msg.guild, 'invitelinkremover');

        if (enabled === 'false') {
            return;
        }

        let result = msg.content.split(' ')
                               .filter((word) => word.length >= this.shortestHostLength)
                               .map((word) => (parseUri(word)).host as string)
                               .filter((host) => this.isBlackListed(host));

        if (result.length === 0) {
            return;
        }

        try {
            await msg.delete();
            await Utils.sendMessage(msg.channel, `I'm sorry ${msg.author}, you can't send invite links here.`);
        } finally {
            return true;
        }
    }

    private isBlackListed(host: string): boolean {
        for (let blacklisted of this.blacklistedHosts) {
            if (host === blacklisted) {
                return true;
            }
        }

        return false;
    }
}

export = InviteLink;

File: safetyjim/inviteLink.go
package safetyjim

import (
	"net/url"
	"strings"

	"github.com/bwmarrin/discordgo"
)

type InviteLink struct {
	BlacklistedHosts    []string
	ShortestsHostLength int
	WhilelistedRoles    []string
}

func NewInviteLink() InviteLink {
	return InviteLink{
		[]string{"discord.gg"}, 10, []string{"ADMINISTRATOR",
			"BAN_MEMBERS",
			"KICK_MEMBERS",
			"MANAGE_ROLES",
			"MANAGE_MESSAGES"},
	}
}

func (i *InviteLink) InviteLinkOnMessage(bot *DiscordBot, s *discordgo.Session, m *discordgo.MessageCreate) bool {
	member := bot.GetMember(s, m.ChannelID, m.Author.ID)
	for _, role := range i.WhilelistedRoles {
		for _, memberRole := range member. {
			if role == memberRole {
				s.ChannelMessageSend(m.ChannelID, "You are exempt from invite link inspection")
				return false
			}
		}
	}

	splitWords := strings.Split(m.Content, " ")

	for _, word := range splitWords {
		if len(word) <= 10 {
			continue
		}

		result, err := url.Parse(word)
		if err != nil {
			continue
		}

		for _, blacklistedHost := range i.BlacklistedHosts {
			if result.Host == blacklistedHost {
				s.ChannelMessageDelete(m.ChannelID, m.ID)
				s.ChannelMessageSend(m.ChannelID, "Bad "+m.Author.Mention()+", you can't send invite links here!")
				return true
			}
		}
	}

	return false
}
